java/ql/src/semmle/code/java/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -3,481 +3,685 @@ import DataFlowImplSpecific::Public
+      exists(DataFlowCall call, ReturnKind kind |
+        storeReturn0(call, kind, node1, f) and
+        node2 = getAnOutNode(call, kind) and
+        compatibleTypes(node1.getTypeBound(), f.getType()) and
+        compatibleTypes(node2.getTypeBound(), f.getContainerType())
+      )
+    }
 
-  /**
-   * Holds if `arg` is a possible argument to `p`, in the call `call`, and
-   * `arg` may flow through `call`. The possible contexts before and after
-   * entering the callable are `outercc` and `innercc`, respectively.
-   */
-  private predicate viableParamArg(
-    DataFlowCall call, ParameterNode p, ArgumentNode arg, CallContext outercc,
-    CallContextCall innercc
-  ) {
-    exists(int i, DataFlowCallable callable | viableParamArg1(p, callable, i, arg, outercc, call) |
-      if reducedViableImplInCallContext(_, callable, call)
-      then innercc = TSpecificCall(call, i, true)
-      else innercc = TSomeCall(p, true)
-    )
-  }
+    private predicate setterReturn(ParameterNode p, Content f, ReturnKind kind) {
+      exists(Node n1, Node n2 |
+        parameterValueFlowNoCtx(p, n1) and
+        store(n1, f, n2) and
+        localValueStep*(n2, getAReturnNodeOfKind(kind))
+      )
+    }
 
-  private CallContextCall getAValidCallContextForParameter(ParameterNode p) {
-    result = TSomeCall(p, _)
-    or
-    exists(DataFlowCall call, int i, DataFlowCallable callable |
-      result = TSpecificCall(call, i, _) and
-      p.isParameterOf(callable, i) and
-      reducedViableImplInCallContext(_, callable, call)
-    )
+    pragma[noinline]
+    private predicate read0(DataFlowCall call, ReturnKind kind, ArgumentNode arg, Content f) {
+      exists(ParameterNode p |
+        viableParamArg(call, p, arg) and
+        getter(p, f, kind)
+      )
+    }
+
+    /**
+     * Holds if data can flow from `node1` to `node2` via a direct read of `f` or
+     * via a getter.
+     */
+    cached
+    predicate read(Node node1, Content f, Node node2) {
+      readStep(node1, f, node2) and storeStep(_, f, _)
+      or
+      exists(DataFlowCall call, ReturnKind kind |
+        read0(call, kind, node1, f) and
+        node2 = getAnOutNode(call, kind) and
+        compatibleTypes(node1.getTypeBound(), f.getContainerType()) and
+        compatibleTypes(node2.getTypeBound(), f.getType())
+      )
+    }
+
+    private predicate getter(ParameterNode p, Content f, ReturnKind kind) {
+      exists(Node n1, Node n2 |
+        parameterValueFlowNoCtx(p, n1) and
+        read(n1, f, n2) and
+        localValueStep*(n2, getAReturnNodeOfKind(kind))
+      )
+    }
+
+    cached
+    predicate localStoreReadStep(Node node1, Node node2) {
+      exists(Node mid1, Node mid2, Content f |
+        store(node1, f, mid1) and
+        localValueStep*(mid1, mid2) and
+        read(mid2, f, node2) and
+        compatibleTypes(node1.getTypeBound(), node2.getTypeBound())
+      )
+    }
+
+    cached
+    module FlowThrough_v2 {
```

Comment:
Perhaps add a comment about how this modules is different from `FlowThrough_v1`.

Reply:
Can you explain why we need two `FlowThrough` modules? AFAICT, they both restrict to value-preserving flow, and v2 adds local flow through fields (`localStoreReadStep`); why is local flow through fields not allowed in the `localValueStep` relation (which is the only use of `FlowThrough_v1`)?

Reply:
The `localValueStep` relation is used to define getters and setters and making it use field flow would turn everything into a big messy SCC.  So the basic dependency order is: Basic flow-through, which is then refined with call contexts (that's `_v1`), use this to define `localValueStep`, use this to define getters and setters, use getters and setters to define the store-localflow-read step as an additional local flow step, with this additional local flow through fields recalculate flow through methods (that's `_v2`).  The reason why we want this additional flow included in the final value-flow-through relation is that the store-localflow-read step is already used in `DataFlowImpl` to calculate taint-flow-through.  This means that until now value-flow-through that makes use of store-localflow-read has been miscategorized as a taint-summary rather than a value-summary.  In follow-up work I intend to fix the taint-flow-through calculation to no longer accidentally include value-flow-through, and that means that without this change we'd lose value-flow-through for those callables that are relying on a store-localflow-read step.

Reply:
I second @hvitved 's request to turn this explanation into code comments.